homeassistant/components/mqtt/device_trigger.py

```diff
@@ -66,23 +66,53 @@
 )
 
 DEVICE_TRIGGERS = "mqtt_device_triggers"
-ATTACHED_DEVICE_TRIGGERS = "mqtt_attached_device_triggers"
+
+
+@attr.s(slots=True)
+class AttachedTrigger:
+    """Attached trigger settings."""
+
+    action = attr.ib(type=AutomationActionType)
+    automation_info = attr.ib(type=dict)
+    remove = attr.ib(type=CALLBACK_TYPE)
 
 
 @attr.s(slots=True)
 class Trigger:
     """Device trigger settings."""
 
+    attached_triggers = attr.ib(type=[AttachedTrigger])
     device_id = attr.ib(type=str)
-    discovery_id = attr.ib(type=str)
     hass = attr.ib(type=HomeAssistantType)
     payload = attr.ib(type=str)
     qos = attr.ib(type=int)
     subtype = attr.ib(type=str)
     topic = attr.ib(type=str)
     type = attr.ib(type=str)
 
-    async def attach_trigger(self, action, automation_info):
+    async def add_trigger(self, action, automation_info):
+        """Add MQTT trigger."""
+        attached_trigger = AttachedTrigger(action, automation_info, None)
```

Comment:
I would call it `TriggerInstance`, since `remove` is not always defined, in which case it's not attached. I think that we should pass in `self` here, so we can access the `config` when we attach inside the `TriggerInstance`

Reply:
Yup, modified!